pkg/codegen/convert/mapper.go

```diff
@@ -14,13 +14,47 @@
 
 package convert
 
-import "context"
+import (
+	"context"
 
-// An interface to map provider names (N.B. These aren't Pulumi provider names, but the names of "providers"
-// in the source language being converted from) to plugin specific mapping data.
+	"github.com/blang/semver"
+)
+
+// Mapper provides methods for retrieving mappings that describe how to map names in some source "provider" (e.g. a
+// Terraform provider, if we are converting from Terraform) to names in appropriate Pulumi packages. So when converting
+// a Terraform program containing code like `resource "aws_s3_bucket" "b" {}`, for instance, we need to know (among
+// other things) that the `aws_s3_bucket` Terraform resource type corresponds to the Pulumi type `aws:s3/bucket:Bucket`,
+// and thus lives in the `aws` package. This is the kind of information that a Mapper provides.
 type Mapper interface {
-	// Returns plugin specific mapping data for the given provider name. The "pulumiProvider" is used as a
-	// hint for which pulumi plugin will provider this mapping. Returns an empty result if no mapping
-	// information was available.
-	GetMapping(ctx context.Context, provider string, pulumiProvider string) ([]byte, error)
+	// GetMapping returns any available mapping data for the given source provider name (so again, this is e.g. the name
+	// of a Terraform provider if converting from Terraform). Callers may pass a "hint" parameter that describes a Pulumi
+	// plugin that is expected to provide the mapping and satisfy the request, which implementations may use to optimise
+	// their efforts to return the best possible mapping. If no matching mapping exists, implementations should return an
+	// empty byte array result.
+	GetMapping(ctx context.Context, provider string, hint *MapperPackageHint) ([]byte, error)
+}
+
+// MapperPackageHint is the type of hints that may be passed to GetMapping to help guide implementations to picking
+// appropriate Pulumi plugins to satisfy mapping requests.
+type MapperPackageHint struct {
+	// The name of the Pulumi plugin that is expected to provide the mapping.
+	PluginName string
+
+	// An optional parameterization that should be used on the named plugin before asking it for mappings.
+	Parameterization *MapperPackageParameterizationHint
```

Comment:
Can this just be a workspace.Parameterization? It's the same shape.

Reply:
Sure